File: src/neural.js
class NeuralNet {
	constructor(layers, act=(a)=>Math.max(a,0), out=(a)=>Math.tanh(a), random = Math.random) {
		if (layers) {
			if (layers.nodes) {
				this.nodes = layers.nodes
				this.paths = layers.paths
				this.act = layers.act
				this.out = layers.out
				this.random = random
			} else {
				this.nodes = layers.map(i => 
					Array(i)
					.fill(0)
					.map((a) => 
						0
						))
				this.act = act
				this.out = out
				this.random = random
				this.paths = []
				for (let layer = 0; layer < layers.length - 1; layer++) {
					let pathLayer = []
					for (let start = 0; start < layers[layer]; start++) {
						for (let end = 0; end < layers[layer + 1]; end++) {
							pathLayer.push([start, end, 1/(layers[layer]*layers[layer + 1])/*layers[layer]*/])
						}
					}
					this.paths.push(pathLayer)
				}
			}

		}
	}
	mutatePaths(amount) {
		if (!this.limit){
			this.paths = this.paths.map(layer => layer
				.map(path => [path[0], path[1], path[2] + ((this.random() - 0.5) * amount)])
				)
		} else {
			this.paths = this.paths.map(layer => layer
				.map(path => {
					let newWeight = path[2] + ((this.random() - 0.5) * amount);
				// Ensure the weights don't grow too large
					newWeight = Math.max(Math.min(newWeight, this.max), -this.max);
					return [path[0], path[1], newWeight];
				})
				);
		}
		console.log(this.paths)
	}

	// Mutation for nodes with limits to prevent large mutations
	mutateNodes(amount) {
		if (!this.max){
			this.nodes = this.nodes.map(layer => layer.map(node => node + ((this.random() - 0.5) * amount)))
		} else {
			this.nodes = this.nodes.map(layer => layer.map(node => {
				let newNode = node + ((this.random() - 0.5) * amount);
			// Ensure nodes don't grow too large
				newNode = Math.max(Math.min(newNode, this.max), -this.max);
				return newNode;
			}));
		}
		console.log(this.nodes)
	}
	clone() {
		let n = new NeuralNet({ nodes: this.nodes.map(l => l.slice()), paths: this.paths.map(l => l.map(p=>p.slice())), act:this.act, out:this.out})
		if (this.max){
			n.max = this.max
		}
		return n
	}
	run(inputs) {
		if (inputs.length !== this.nodes[0].length) {
			throw new TypeError("input size incorrect: " + inputs.size)
		}
		let model = this.clone()
		inputs.forEach((i, j) => model.nodes[0][j] = i)
		model.paths.forEach((layer, layerNum) => {
			layer.forEach(path => {
				model.nodes[layerNum + 1][path[0]] += model.nodes[layerNum][path[1]] * path[2]
			})
			model.nodes[layerNum+1] = model.nodes[layerNum+1].map(a=>this.act(a))
		})
		return model.nodes.pop().map(a=>this.out(a))
	}

	toString() {
		return JSON.stringify({ nodes: this.nodes.map(l => l.map(v=>parseFloat(v.toFixed(10)))), paths: this.paths.map(l => l.map(p=>{return [p[0], p[1], parseFloat(p[2].toFixed(10))]})), act:this.act.toString(), out:this.out.toString()})
	}
	fromString(str) {
		let m = JSON.parse(str)
		m.act = eval("(()=>{return " + m.act + "})()")
		m.out = eval("(()=>{return " + m.out + "})()")
		return new NeuralNet(m)
	}
	from(net){
		return net.clone()
	}
	draw(ctx, size = 20) {
		let normalised = this.clone()
		let maxNode = Math.max(...normalised.nodes.flat());
		let maxPath = Math.max(...normalised.paths.map(l=>l.map(p => p[2])).flat());
		normalised.nodes = normalised.nodes.map(l => l.map(n => (n / maxNode) * size));
		normalised.paths.forEach(layer => layer.forEach(path => path[2] = (path[2] / maxPath) * size/4));

		const layerWidth = (ctx.canvas.width - 80) / (normalised.nodes.length + 1);
		const radius = 20; // Radius of the nodes

		let nodePositions = [];

		// Draw the nodes
		for (let layer = 0; layer < normalised.nodes.length; layer++) {
			const layerHeight = (ctx.canvas.height - 80) / (normalised.nodes[layer].length + 1);
			let currentLayerPositions = [];

			for (let node = 0; node < normalised.nodes[layer].length; node++) {
				const x = ((layer + 1) * layerWidth) + 40;
				const y = ((node + 1) * layerHeight) + 40;
				currentLayerPositions.push({ x, y });

				// Draw the node
				ctx.lineWidth = 2
				ctx.beginPath();
				ctx.arc(x, y, Math.abs(normalised.nodes[layer][node]), 0, Math.PI * 2);
				ctx.fillStyle = "#3498db";
				ctx.fill();
				ctx.stroke();
			}

			nodePositions.push(currentLayerPositions);
		}

		// Draw the paths (connections)
		ctx.strokeStyle = "#2c3e50";
		for (let layer = 0; layer < normalised.paths.length; layer++) {
			for (let path of normalised.paths[layer]) {
				const start = nodePositions[layer][path[0]];
				const end = nodePositions[layer + 1][path[1]];
				ctx.lineWidth = Math.abs(path[2])

				ctx.beginPath();
				ctx.moveTo(start.x, start.y);
				ctx.lineTo(end.x, end.y);
				ctx.stroke();
			}
		}
	}
}


if (typeof self == "undefined"){
	module.exports = {NeuralNet}
}

File: src/neural.py
import json
import random
import math

class NeuralNet:
    def __init__(self, layers, act=lambda a: max(a, 0), out=math.tanh, random_gen=random.random):
        self.act = act
        self.out = out
        self.random = random_gen
        
        if isinstance(layers, dict) and 'nodes' in layers:
            self.nodes = layers['nodes']
            self.paths = layers['paths']
        else:
            self.nodes = [[0 for _ in range(i)] for i in layers]
            self.paths = []
            for layer in range(len(layers) - 1):
                path_layer = [
                    [start, end, 1 / (layers[layer] * layers[layer + 1])]
                    for start in range(layers[layer])
                    for end in range(layers[layer + 1])
                ]
                self.paths.append(path_layer)

    def mutate_paths(self, amount):
        if not hasattr(self, 'max'):
            self.paths = [
                [[start, end, weight + (self.random() - 0.5) * amount] for start, end, weight in layer]
                for layer in self.paths
            ]
        else:
            self.paths = [
                [[start, end, max(min(weight + (self.random() - 0.5) * amount, self.max), -self.max)]
                 for start, end, weight in layer]
                for layer in self.paths
            ]

    def mutate_nodes(self, amount):
        if not hasattr(self, 'max'):
            self.nodes = [
                [node + (self.random() - 0.5) * amount for node in layer]
                for layer in self.nodes
            ]
        else:
            self.nodes = [
                [max(min(node + (self.random() - 0.5) * amount, self.max), -self.max) for node in layer]
                for layer in self.nodes
            ]

    def clone(self):
        cloned_net = NeuralNet(
            {
                'nodes': [layer[:] for layer in self.nodes],
                'paths': [[path[:] for path in layer] for layer in self.paths],
                'act': self.act,
                'out': self.out
            }
        )
        if hasattr(self, 'max'):
            cloned_net.max = self.max
        return cloned_net

    def run(self, inputs):
        if len(inputs) != len(self.nodes[0]):
            raise TypeError(f"input size incorrect: {len(inputs)}")

        model = self.clone()
        model.nodes[0] = inputs[:]
        
        for layer_num, layer in enumerate(model.paths):
            for start, end, weight in layer:
                model.nodes[layer_num + 1][end] += model.nodes[layer_num][start] * weight
            model.nodes[layer_num + 1] = [self.act(a) for a in model.nodes[layer_num + 1]]

        return [self.out(a) for a in model.nodes[-1]]

    def __str__(self):
        return json.dumps({
            'nodes': [[round(v, 10) for v in layer] for layer in self.nodes],
            'paths': [[[start, end, round(weight, 10)] for start, end, weight in layer] for layer in self.paths],
            'act': self.act.__code__.co_code.decode("utf-8"),
            'out': self.out.__code__.co_code.decode("utf-8")
        })

    @staticmethod
    def from_string(data_str):
        data = json.loads(data_str)
        act = eval(f"lambda a: {data['act']}")
        out = eval(f"lambda a: {data['out']}")
        net = NeuralNet({'nodes': data['nodes'], 'paths': data['paths'], 'act': act, 'out': out})
        return net

    def from_net(self, net):
        return net.clone()

# Usage
if __name__ == "__main__":
    net = NeuralNet([3, 2, 1])
    print(net)